### PR TITLE
godeps: replace "form3tech-oss/jwt-go" with "golang-jwt/jwt"   

### LINKED ISSUE #1871: Replace github.com/form3tech-oss/jwt-go with https://github.com/golang-jwt/jwt
### Observed behavior
We have been using github.com/form3tech-oss/jwt-go, which now should be replaced with https://github.com/golang-jwt/jwt

### Details on how to reproduce (minimal and precise)
Since the maintained fork of https://github.com/form3tech-oss/jwt-go seems now to be https://github.com/golang-jwt/jwt, we might need to make some changes accordingly.  For more information on this please refer to kubernetes/kubernetes#104007

### Information about the environment:

- Heketi version used (e.g. v6.0.0 or master): 
- Operating system used: 
- Heketi compiled from sources, as a package (rpm/deb), or container: 
- If container, which container image: 
- Using kubernetes, openshift, or direct install: 
- If kubernetes/openshift, is gluster running inside kubernetes/openshift or outside: 
- If kubernetes/openshift, how was it deployed (gk-deploy, openshift-ansible, other, custom):

### PR DESCRIPTION
**Pull request by Haimantika**
use a community maintained fork of jwt
for reference: kubernetes/kubernetes#104007

Signed-off-by: haimantika mitra <haimantikamitra@gmail.com>

### Does this PR fix issues?

Fixes #

### CODE REVIEW

**Comment by Haimantika** [comment]
> Can one of the admins verify this patch?

Yeah it will be great then!

**Comment by phlogistonjohn** [comment]
On first pass the PR looks OK. However, we were already in the midst of an effort to move to go modules such that heketi will be continue to be build-able on future versions of Go where only modules are supported. As such, could I please bother you to either rebase your patches on top of the work in PR #1867  or just hang in there until we've finalized go modules and merged the PR?

Thanks!

**Comment by Haimantika** [comment]
> On first pass the PR looks OK. However, we were already in the midst of an effort to move to go modules such that heketi will be continue to be build-able on future versions of Go where only modules are supported. As such, could I please bother you to either rebase your patches on top of the work in PR #1867 or just hang in there until we've finalized go modules and merged the PR?
> 
> Thanks!

Both of it is do-able, however I feel the later is more suitable? I am a beginner in this, so whatever you suggest is better, we can go ahead with it?

**Comment by phlogistonjohn** [comment]
> Both of it is do-able, however I feel the later is more suitable? I am a beginner in this, so whatever you suggest is better, we can go ahead with it?

Agreed. If no one is in a big rush the latter is certainly the simplest and easiest option. I suggest subscribing to PR #1867 so you can be alerted to changes on it but I'll also try to drop a reminder here. Thank you.

**Comment by iamniting** [comment]
@Haimantika Can you also update the commit something like below?

```
godeps: replace "form3tech-oss/jwt-go" with "golang-jwt/jwt"

use a community maintained fork of jwt
for reference: https://github.com/kubernetes/kubernetes/issues/104007
```

**Comment by Haimantika** [comment]
> @Haimantika Can you also update the commit something like below?
> 
> ```
> godeps: replace "form3tech-oss/jwt-go" with "golang-jwt/jwt"
> 
> use a community maintained fork of jwt
> for reference: https://github.com/kubernetes/kubernetes/issues/104007
> ```

Hey! Do you mean the commit message?

**Comment by iamniting** [comment]
> > @Haimantika Can you also update the commit something like below?
> > ```
> > godeps: replace "form3tech-oss/jwt-go" with "golang-jwt/jwt"
> > 
> > use a community maintained fork of jwt
> > for reference: https://github.com/kubernetes/kubernetes/issues/104007
> > ```
> 
> Hey! Do you mean the commit message?

yes

**Comment by Haimantika** [comment]
> > > @Haimantika Can you also update the commit something like below?
> > > ```
> > > godeps: replace "form3tech-oss/jwt-go" with "golang-jwt/jwt"
> > > 
> > > use a community maintained fork of jwt
> > > for reference: https://github.com/kubernetes/kubernetes/issues/104007
> > > ```
> > 
> > 
> > Hey! Do you mean the commit message?
> 
> yes

Just changed, does it look fine now?

**Comment by iamniting** [comment]
> > > > @Haimantika Can you also update the commit something like below?
> > > > ```
> > > > godeps: replace "form3tech-oss/jwt-go" with "golang-jwt/jwt"
> > > > 
> > > > use a community maintained fork of jwt
> > > > for reference: https://github.com/kubernetes/kubernetes/issues/104007
> > > > ```
> > > 
> > > 
> > > Hey! Do you mean the commit message?
> > 
> > 
> > yes
> 
> Just changed, does it look fine now?

Not really, the header of your commit-msg is too long, it should exactly look like mines. Do not repeat the body twice in the header and body.
Screenshot from 2021-08-05 13-47-21
.

**Comment by Haimantika** [comment]
> > > > > @Haimantika Can you also update the commit something like below?
> > > > > ```
> > > > > godeps: replace "form3tech-oss/jwt-go" with "golang-jwt/jwt"
> > > > > 
> > > > > use a community maintained fork of jwt
> > > > > for reference: https://github.com/kubernetes/kubernetes/issues/104007
> > > > > ```
> > > > 
> > > > 
> > > > Hey! Do you mean the commit message?
> > > 
> > > 
> > > yes
> > 
> > 
> > Just changed, does it look fine now?
> 
> Not really, the header of your commit-msg is too long, it should exactly look like mines. Do not repeat the body twice in the header and body.
> Screenshot from 2021-08-05 13-47-21

Oops! My bad! Fixed it now!

**Comment by iamniting** [comment]
The commit-msg looks good to me now. As https://github.com/heketi/heketi/pull/1867 is merged you can rebase and make appropriate changes. We will CI after that.

**Comment by Haimantika** [comment]
> The commit-msg looks good to me now. As #1867 is merged you can rebase and make appropriate changes. We will run CI after that.

I did this yesterday itself.

**Comment by iamniting** [comment]
> > The commit-msg looks good to me now. As #1867 is merged you can rebase and make appropriate changes. We will run CI after that.
> 
> I did this yesterday itself.

We no longer need glide.yaml, moreover you need to make changes to go.mod and go.sum

**Comment by Haimantika** [comment]
> > > The commit-msg looks good to me now. As #1867 is merged you can rebase and make appropriate changes. We will run CI after that.
> > 
> > 
> > I did this yesterday itself.
> 
> We no longer need glide.yaml, moreover you need to make changes to go.mod and go.sum

Yeah. The glide.yaml file I removed. Looks like it did not work! Let me try again and update you here.

**Comment by Haimantika** [comment]
> > > The commit-msg looks good to me now. As #1867 is merged you can rebase and make appropriate changes. We will run CI after that.
> > 
> > 
> > I did this yesterday itself.
> 
> We no longer need glide.yaml, moreover you need to make changes to go.mod and go.sum

Does it look good now?

However why am I seeing this? Will this affect in any way?
Capture

**Comment by iamniting** [comment]
> > > > The commit-msg looks good to me now. As #1867 is merged you can rebase and make appropriate changes. We will run CI after that.
> > > 
> > > 
> > > I did this yesterday itself.
> > 
> > 
> > We no longer need glide.yaml, moreover you need to make changes to go.mod and go.sum
> 
> Does it look good now?
> 
> However why am I seeing this? Will this affect in any way?
> Capture

It means your branch is not rebased with the base. you can do this again after pulling the master branch

It is a little better than before however, you updated the commit-msg again and I was expecting something to be deleted from the go.sum also

Screenshot from 2021-08-05 14-53-04

**Comment by Haimantika** [comment]
> > > > > The commit-msg looks good to me now. As #1867 is merged you can rebase and make appropriate changes. We will run CI after that.
> > > > 
> > > > 
> > > > I did this yesterday itself.
> > > 
> > > 
> > > We no longer need glide.yaml, moreover you need to make changes to go.mod and go.sum
> > 
> > 
> > Does it look good now?
> > However why am I seeing this? Will this affect in any way?
> > Capture
> 
> It means your branch is not rebased with the base. you can do this again after pulling the master branch
> 
> It is a little better than before however, you updated the commit-msg again and I was expecting something to be deleted from the go.sum also. you can try to run go mod tidy
> 
> Screenshot from 2021-08-05 14-53-04

Oh yes. That is after the commit I made, will ammend it. The go.sum file is such after running go mod tidy.

**Comment by iamniting** [comment]
@Haimantika Can you pls correct the changes so we can take this to closure?

**Comment by Haimantika** [comment]
> @Haimantika Can you pls correct the changes so we can take this to closure?

I accidently removed the main.go file from the GitHub UI, which is causing a conflict and I am not able to resolve it.

**Comment by iamniting** [comment]
> > @Haimantika Can you pls correct the changes so we can take this to closure?
> 
> I accidently removed the main.go file from the GitHub UI, which is causing a conflict and I am not able to resolve it.

You can do

`git reset --soft HEAD~1`
`git reset main.go`
`git checkout main.go`

and run `git commit` after it to commit the changes

**Comment by Haimantika** [comment]
> > > @Haimantika Can you pls correct the changes so we can take this to closure?
> > 
> > 
> > I accidently removed the main.go file from the GitHub UI, which is causing a conflict and I am not able to resolve it.
> 
> You can do
> 
> `git reset --soft HEAD~1`
> `git reset main.go`
> `git checkout main.go`
> 
> and run `git commit` after it to commit the changes

Thanks! It is done. The second commit has the required changes.

**Comment by iamniting** [comment]
> > > > @Haimantika Can you pls correct the changes so we can take this to closure?
> > > 
> > > 
> > > I accidently removed the main.go file from the GitHub UI, which is causing a conflict and I am not able to resolve it.
> > 
> > 
> > You can do
> > `git reset --soft HEAD~1`
> > `git reset main.go`
> > `git checkout main.go`
> > and run `git commit` after it to commit the changes
> 
> Thanks! It is done. The second commit has the required changes.

We can not just merge the 2nd commit, we need to merge the whole PR. you need to squash them.

**Comment by Haimantika** [comment]
> > > > > @Haimantika Can you pls correct the changes so we can take this to closure?
> > > > 
> > > > 
> > > > I accidently removed the main.go file from the GitHub UI, which is causing a conflict and I am not able to resolve it.
> > > 
> > > 
> > > You can do
> > > `git reset --soft HEAD~1`
> > > `git reset main.go`
> > > `git checkout main.go`
> > > and run `git commit` after it to commit the changes
> > 
> > 
> > Thanks! It is done. The second commit has the required changes.
> 
> We can not just merge the 2nd commit, we need to merge the whole PR. you need to squas

> > > > > @Haimantika Can you pls correct the changes so we can take this to closure?
> > > > 
> > > > 
> > > > I accidently removed the main.go file from the GitHub UI, which is causing a conflict and I am not able to resolve it.
> > > 
> > > 
> > > You can do
> > > `git reset --soft HEAD~1`
> > > `git reset main.go`
> > > `git checkout main.go`
> > > and run `git commit` after it to commit the changes
> > 
> > 
> > Thanks! It is done. The second commit has the required changes.
> 
> We can not just merge the 2nd commit, we need to merge the whole PR. you need to squash them.

Have one commit now, but we have a Dockerfile.

**Comment by humblec** [comment]
@iamniting jfyi, the  docker build on `master` looks to be broken btw after the commit which removed `glide.yaml` .

**Comment by iamniting** [comment]
> @iamniting jfyi, the docker build on `master` looks to be broken btw after the commit which removed `glide.yaml` .

@humblec Can you be more specific about which docker build?

File: docs/profiling.md [hunk 1]
```diff
@@ -63,7 +63,7 @@ Showing top 10 nodes out of 93
     6.50MB  3.25% 77.23%     6.50MB  3.25%  context.WithValue
     6.50MB  3.25% 80.48%     6.50MB  3.25%  encoding/json.(*decodeState).literalStore
        6MB  3.00% 83.49%     6.50MB  3.25%  net/textproto.(*Reader).ReadLine
-    5.50MB  2.75% 86.24%    39.51MB 19.76%  github.com/heketi/heketi/vendor/github.com/form3tech-oss/jwt-go.(*Parser).ParseWithClaims
+    5.50MB  2.75% 86.24%    39.51MB 19.76%  github.com/heketi/heketi/vendor/github.com/golang-jwt/jwt-go.(*Parser).ParseWithClaims
```
Comment (iamniting) [inline docs/profiling.md:66]:
```suggestion
    5.50MB  2.75% 86.24%    39.51MB 19.76%  github.com/heketi/heketi/vendor/github.com/golang-jwt/jwt.(*Parser).ParseWithClaims
```

**Review by iamniting — CHANGES_REQUESTED** [review]

**Comment by iamniting** [comment]
@phlogistonjohn The PR is ready now Can you pls run the action?

**Review by iamniting — APPROVED** [review]

**Comment by iamniting** [comment]
I think we can remove the older versions from the action now as they are not fully supported with the `golang-jwt/jwt`

**Comment by phlogistonjohn** [comment]
> I think we can remove the older versions from the action now as they are not fully supported with the `golang-jwt/jwt`

If the new jwt library breaks on older go versions I am very very reluctant to use it. I *know* that heketi is used on older systems/distributions.

**Comment by iamniting** [comment]
> > I think we can remove the older versions from the action now as they are not fully supported with the `golang-jwt/jwt`
> 
> If the new jwt library breaks on older go versions I am very very reluctant to use it. I _know_ that heketi is used on older systems/distributions.

agreed

**Comment by dims** [comment]
@phlogistonjohn unfortunately form3tech-oss/jwt-go is out of business :(

image

https://endoflife.date/go shows golang 1.2.x/1.3.x/1.4.x all out of support. 1.4.x went away about 18 months ago. Does heketi need support all the way back to 1.2? we could investigate that, but seems daunting task.

one obvious thing that showed up was that support for Ed25519 is absent in `form3tech-oss/jwt-go` and was added in `golang-jwt/jwt`

**Comment by phlogistonjohn** [comment]
> @phlogistonjohn unfortunately form3tech-oss/jwt-go is out of business :( - https://github.com/form3tech-oss/jwt-go shows
> 
> image
> 

OK, that's something for sure. In my defense I do believe that was changed recently, after this PR was filed. :-)



> https://endoflife.date/go shows golang 1.12.x/1.13.x/1.14.x all out of support. 1.14.x went away about 18 months ago. 

Yes, that's right and we're aware of it.


> Does heketi need support all the way back to 1.12? we could investigate that, but seems daunting task.
> 

We know we support users on centos 7.  I've recently responded to users complaining about issues related to centos 7. Now, its true that one could "install" Go from the tarball releases, but we're also aware that not everyone can just pull stuff from the internet. So I am a bit concerned about cutting them off.

> one obvious thing that showed up was that support for Ed25519 is absent in `form3tech-oss/jwt-go` and was added in `golang-jwt/jwt`

Yeah, that looks like the main thing breaking the build on the older versions.

One of the main drivers of any of the work we do on heketi is security. So there's a clear impetus to take these new versions of Go and jwt library, it just worries me that the language/toolchain and library don't sync well with our legacy use case here in heketi.

Since users on centos 7 and the like can still build heketi with non-packaged versions of Go, use heketi from a container, or take a heketi binary we provide (and we plan on fixing issues wrt that for the next release) I don't think this is a change we should  neglect, but I think I've also made it clear that this isn't a trivial decision.

Thanks for your feedback.

**Comment by phlogistonjohn** [comment]
We've removed the older go versions from the CI. Please rebase again, thanks.

**Review by raghavendra-talur — APPROVED** [review]

**Review by phlogistonjohn — APPROVED** [review]
Tests are now passing, changes are acceptable (see previous discussions), I think we ought to take this.

**Comment by humblec** [comment]
> Tests are now passing, changes are acceptable (see previous discussions), I think we ought to take this.

Awesome..  Thanks !
